### PR TITLE
PLT-1465 Added password requirements

### PR DESCRIPTION
**Pull request by DavidLu1997**
Please make sure you've read the [pull request](http://docs.mattermost.com/developer/contribution-guide.html#preparing-a-pull-request) section of our [code contribution guidelines](http://docs.mattermost.com/developer/contribution-guide.html).

#### Summary
Passwords are now validated server and client side using requirements set by the system admin.

Possible requirements:

Minimum length (default 5)

At least one lowercase letter
At least one uppercase letter
At least one number
At least one symbol (almost all symbols supported)
This should work for everywhere that a password change is required, existing passwords are untouched.

All password settings are EE only, the system defaults to enforcing a minimum length of 5 without a license.

#### Unit Tests
A lot of them

#### Issue/Ticket Link
https://mattermost.atlassian.net/browse/PLT-1465

#### Has Critical Changes
- Authentication
- Passwords
- Log in
- Sign Up

### CODE REVIEW

**Comment by jwilander** [comment]
+1

**Comment by DavidLu1997** [comment]
@jwilander rebased

**Comment by esethna** [comment]
@crspeller the spinmint above doesn't seem to have enterprise enabled?

**Comment by crspeller** [comment]
@esethna It's certainly enterprise enabled. Just upload a licence.

**Comment by esethna** [comment]
@DavidLu1997, can we tweak:

1. The help text below minimum password length: "Minimum number of characters required for a valid password. Must be a whole number greater or equal to 5 and less than or equal to 64."
2. If you enter a min password length less than 3, the error message previews even though the input is invalid: 
image

**Comment by esethna** [comment]
+1 other than that, David and I tested together:

1. Putting invalid inputs into the min password length
3. Verified the preview in the system console matches the error messages on the account creation page
4. Verified the preview in the system console matches the error messages on the change password page
4. Verified that old passwords are still valid even if the requirements are made more strict

**Comment by DavidLu1997** [comment]
Tweaks added:
- Changed help text as above
- Enforced minimum length of preview
- Moved Password Reset Salt above Login Attempts and MFA

File: api/user.go [hunk 1]
```diff
@@ -245,6 +245,11 @@ func CreateUser(user *model.User) (*model.User, *model.AppError) {
 	user.MakeNonNil()
 	user.Locale = *utils.Cfg.LocalizationSettings.DefaultClientLocale
 
+	if err := utils.IsPasswordValid(user.Password); user.AuthService == "" && err != nil {
+		err.StatusCode = http.StatusNotAcceptable
```
Comment (coreyhulen) [inline api/user.go]: Not sure we use this status code anywhere else?  I think 500 or invalid arg is good.

**Comment by coreyhulen** [comment]
minor nit about error code otherwise +1

**Comment by coreyhulen** [comment]
needs a rebase

**Comment by coreyhulen** [comment]
@DavidLu1997 needs a rebase

**Comment by DavidLu1997** [comment]
@coreyhulen rebased, should I make a commit to fix the error code?

**Comment by coreyhulen** [comment]
@DavidLu1997 yes, please fix the error code.  I don't think we use it anywhere else.

**Comment by DavidLu1997** [comment]
@coreyhulen changed error code to 500

**Comment by coreyhulen** [comment]
@DavidLu1997 can we change all the StatusNotAcceptable to 500's.  Also the default in AppError is 500 so you shouldn't need to set it.

**Comment by DavidLu1997** [comment]
@coreyhulen removed all error code setting

**Pull request merged**